[SelectionDAG] Add support to filter SelectionDAG dumps during ISel by function names

mshockwave opened the pull request:
`-debug-only=isel-dump` is the new debug type for printing SelectionDAG after each ISel phase. This can be furthered filter by `-filter-print-funcs=<function names>`.

mshockwave committed:
[SelectionDAG] Add a flag to filter ISel traces by function name

With `-filter-print-dags-funcs=<function names>`, `-debug-only=isel`
will only print out traces that match the given function names.

mshockwave committed:
fixup! [SelectionDAG] Add a flag to filter ISel traces by function name

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:
```diff
@@ -142,6 +142,12 @@ UseMBPI("use-mbpi",
 #ifndef NDEBUG
+static cl::list<std::string> FilterDAGFuncNames(
+    "filter-print-dags-funcs", cl::Hidden, cl::value_desc("function names"),
+    cl::desc("Only print DAGs of functions whose names "
```
I think we should take a step back and figure out what the goal of `-debug-only=isel -filter-print-dags-funcs=` is.

This patch seems to have touched every debug message in this file, but this isn't the only SelectionDAG file.

Are you just wanting to see the dumps of the DAG from  `CodeGenAndEmitDAG()` for a specific function? Do you care about any of the other messages for that function? Its not clear that you'd only want those debug messages in this file and not the other SelectionDAG files. Since the split between files isn't completely clean.

If you only want to see the dumps from `CodeGenAndEmitDAG()`, then maybe those should be given their own DEBUG_TYPE separate from the rest of this file. So you could do `-debug-only=isel-dumps -filter-print-dags-funcs=`

If you want all debug messages for a particular function then maybe this property should live on SelectionDAG and filter all of the messages in all of the SelectionDAG related files?

ilovepi replied:
Seems like this should have a scope very similar to `FilterDAGBasicBlockName`. I'm also a bit confused why we need an new option, since it feels very similar to `-filter-print-funcs=`... can we just reuse that? IMO asking for a dump of the DAG + `-filter-print-funcs=` seems natural, but maybe it's missing the desired granularity, or would it include information you'd like to exclude from the output?

mshockwave replied:
> If you only want to see the dumps from CodeGenAndEmitDAG(), then maybe those should be given their own DEBUG_TYPE separate from the rest of this file. So you could do -debug-only=isel-dumps -filter-print-dags-funcs=

Printing traces in CodeGenAndEmitDAG is my primary goal, so I'm probably going with the `-debug-only=isel-dumps` + filter function flag approach.

> it feels very similar to -filter-print-funcs=... can we just reuse that? IMO asking for a dump of the DAG + -filter-print-funcs= seems natural

Sounds like a good idea to me. I just found that several CodeGen Passes are also using `-filter-print-funcs`.

> but maybe it's missing the desired granularity, or would it include information you'd like to exclude from the output?

I don't think these will be a problem.

mshockwave committed:
Simplify the filtering logics

  - Re-use `-filter-print-funcs=<function names>`
  - Create a separate debug type, "isel-dump", for DAG dumps to be
  filtered by function names. Note the the original "isel" debug type
  still works the same.

mshockwave commented:
I've made some major revision to this patch:
  1. Re-use `-filter-print-funcs=<function names>`.
  2. Introducing a new debug type, "isel-dump", for printing SelectionDAG after each ISel phase. This debug type can be filtered by `-filter-print-funcs`.  Note that it is also compatible with the original "isel" type. Namely, `-debug-only=isel` will still print the full, unfiltered debug output regardless the value of `-filter-print-funcs`.

ilovepi commented:
Can you document the new flag and behavior? This seems like something useful to communicate. I guess we should probably mention the change of behavior in the release notes too.

ilovepi commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:
```diff
@@ -180,6 +182,15 @@ static const bool ViewDAGCombine1 = false, ViewLegalizeTypesDAGs = false,
 
+#define ISEL_DUMP(F, X)                                                        \
+  do {                                                                         \
+    if (llvm::DebugFlag && (isCurrentDebugType(DEBUG_TYPE) ||                  \
+                            (isCurrentDebugType(ISEL_DUMP_DEBUG_TYPE) &&       \
+                             isFunctionInPrintList(F)))) {                     \
```
This seems a bit heavy to check at every site, right? I think we could at least cache `isFunctionInPrintList(F)` at the start of the pass.  Off the top of my head, I can't think of a nicer way to check the `DEBUG_TYPE` in the current implementation, but maybe alternate structure would help?

Maybe I'm overthinking this detail though, and perf isn't so important if you're debug printing something ...

mshockwave replied:
Done. Now it stores the filter result in a class member variable of `SelectionDAGISel`. Interesting side note: `isCurrentDebugType`, which is used by `LLVM_DUMP`, actually does a linear search to filter debug types every time it's called. But I guess the size of `-debug-only` is not big so it's fine for majority of the time.

mshockwave committed:
Address code reivew feedbacks

  - Store function filtering result as a class member variable instead
  of recalculate in each ISEL_DUMP.
  - Add this feature into the release note.

mshockwave commented:
> Can you document the new flag and behavior? This seems like something useful to communicate. I guess we should probably mention the change of behavior in the release notes too.

I haven't found a proper `docs/*.rst` to put this thing yet but I've put it in the release notes.

mshockwave committed:
fixup! Address code reivew feedbacks

ilovepi commented:
> I haven't found a proper `docs/*.rst` to put this thing yet but I've put it in the release notes.

Maybe `docs/CodeGenerator.rst`? there is a section on instruction selection that lists some debug options, like `-view-dag-combine1-dags` and `-filter-view-dags`, but IDK if there is a more appropriate spot in another file. Nothing else immediately jumps out at me, but maybe @topperc has a better suggestion?

ilovepi commented on llvm/docs/ReleaseNotes.rst:
```diff
@@ -201,6 +201,14 @@ Changes to the C API
 -------------------------------------
 
+* A new debug type ``isel-dump`` is added to show only the SelectionDAG dumps
+  after each ISel phase (i.e. ``-debug-onlu=isel-dump``). This new debug type
```
typo: `-debug-only=`

mshockwave replied:
Done.

ilovepi commented on llvm/docs/ReleaseNotes.rst:
```diff
@@ -201,6 +201,14 @@ Changes to the C API
+  can be filtered by function names using ``-filter-print-funcs=<function names>``,
+  the same flag used to filter IR dumps after each Pass. Note that to be
+  compatible with the existing ``-debug-only=isel``, the latter will still
+  print SelectionDAG dumps of every single functions regardless of
```
s/functions/function

mshockwave replied:
Done

ilovepi commented on llvm/docs/ReleaseNotes.rst:
```diff
@@ -201,6 +201,14 @@ Changes to the C API
+  can be filtered by function names using ``-filter-print-funcs=<function names>``,
+  the same flag used to filter IR dumps after each Pass. Note that to be
+  compatible with the existing ``-debug-only=isel``, the latter will still
+  print SelectionDAG dumps of every single functions regardless of
```
I think this sentence is a bit confusing to read. For example, I'm not sure what `the latter` refers to in this context.

Do you think this change would still make sense?

```
Note that -debug-only=isel will take precedence over the new behavior and print SelectionDAG dumps ...
```

mshockwave replied:
Yes, that's better. It's fixed now.

ilovepi commented on llvm/include/llvm/CodeGen/SelectionDAGISel.h:
```diff
@@ -61,6 +61,10 @@ class SelectionDAGISel : public MachineFunctionPass {
   std::unique_ptr<OptimizationRemarkEmitter> ORE;
 
+  /// True if the function currently processing is in the function printing list
+  /// (i.e. `-filter-print-funcs`).
```
Probably worth noting here (or in the macro?) that we want to cache this value so we don't have to query the module and scan the list at every print? seems like a detail that would be easy to miss later.

mshockwave replied:
Done.

ilovepi reviewed (approved):
I think this patch is good once the rather minor nits and documentation have been addressed.

I'd double check w/ @topperc before landing, though, since he had some questions on the overall direction.

mshockwave committed:
Minor fixes in documentation and release notes

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:
```diff
@@ -180,6 +182,15 @@ static const bool ViewDAGCombine1 = false, ViewLegalizeTypesDAGs = false,
 #endif
 
+#define ISEL_DUMP(X)                                                           \
+  do {                                                                         \
```
Do we need to define this to an empty block when NDEBUG is defined? Otherwise I think we try to evaluate `X` but the `FuncName` variable passed to the first ISEL_DUMP won't be defined.

mshockwave replied:
Oops, I thought I passed `X` into `LLVM_DEBUG` (which will not evaluate X in NDEBUG). It's fixed now.

mshockwave committed:
Add documentions into docs/CodeGenerator.rst

mshockwave committed:
fixup! Address code reivew feedbacks

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:
```diff
@@ -180,6 +182,21 @@ static const bool ViewDAGCombine1 = false, ViewLegalizeTypesDAGs = false,
+    }                                                                          \
+  } while (false)
+#else
+#define ISEL_DUMP(X)                                                           \
```
Just put the whole macro on one line like Debug.h

`#define ISEL_DUMP(X) do { } while (false)`

mshockwave replied:
clang-format somehow disagrees with this.

mshockwave replied:
Well, since we don't need to pass the formatting check to commit this change, I'll just make it one-line.

mshockwave committed:
fixup! Address code reivew feedbacks

topperc reviewed (approved):
LGTM
